src/hash/hasher.cr

```diff
@@ -0,0 +1,188 @@
+require "crystal/system/random"
+
+# Hasher usable for `def hash(hasher)` should satisfy protocol:
+# ```
+# class MyHasher
+#   # Value should implement commutative `+` for `Hash#hash(hasher)`
+#   alias Value
+#
+#   # must be implemented to mix sizes of collections, and pointers (object_id)
+#   def raw(v : Int::Primitive)
+#     # mutate
+#     nil
+#   end
+#
+#   # must be implemented for Hash#hash
+#   def raw(v : Value)
+#     # mutate
+#     nil
+#   end
+#
+#   def <<(b : Bytes)
+#     # mutate
+#     nil
+#   end
+#
+#   def <<(n : Nil)
+#     # mutate
+#     nil
+#   end
+#
+#   def <<(v)
+#     # v.hash will return hasher
+#     # if hasher is a struct, then it will be copy
+#     copy_from v.hash(self)
+#     nil
+#   end
+#
+#   # digest returns hashsum for current state without state mutation
+#   def digest : Value
+#   end
+#
+#   # should be implemented for `Hash#hash(hasher)`
+#   def clone
+#     copy_of_current_state
+#   end
+# end
+# ```
+
+class Hash
+  # Hasher used as standard hasher in `Object#hash`
+  struct Hasher
+    # Type for Hash::Hasher#digest
+    alias Value = UInt32
+
+    @@seed = uninitialized StaticArray(UInt32, 1)
+    buf = pointerof(@@seed).as(Pointer(UInt8))
+    Crystal::System::Random.random_bytes(buf.to_slice(sizeof(typeof(@@seed))))
+
+    protected getter a : UInt32 = 0_u32
+
+    # Construct hasher with per-process seed.
+    def initialize
+      @a = @@seed[0]
+    end
+
+    # Construct hasher with custom seed.
+    def initialize(@a : UInt32)
+    end
+
+    # Calculate hashsum for value
+    def self.hashit(value) : Value
+      s = new(@@seed[0])
+      s << value
+      s.digest
+    end
+
+    # Make copy of self.
+    #
+    # Primary usage is for `Hash#hash` calculation.
+    def clone : self
+      self.class.new(@a)
+    end
+
+    # Mix nil to state
+    def <<(v : Nil) : Nil
```

Comment:
The only function of these methods seem to be calling `permute`. Why not move the permute body into this function?